File: DSA/LC/36_valid_sudoku.py
import collections
import math
from typing import List


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = collections.defaultdict(set)
        cols = collections.defaultdict(set)
        squares = collections.defaultdict(set)

        for i in range(9):
            for j in range(9):
                cell = board[i][j]

                if cell == '.':
                    continue

                square_index = math.floor(i / 3) * 3 + math.floor(j / 3)

                if (
                        cell in rows[i]
                        or cell in cols[j]
                        or cell in squares[square_index]
                ):
                    return False

                rows[i].add(cell)
                cols[j].add(cell)
                squares[square_index].add(cell)
        return True

File: DSA/LC/36_valid_sudoku.js
function isValidSudoku(board) {
  let rows = Array.from({ length: 9 }, () => new Set());
  let cols = Array.from({ length: 9 }, () => new Set());
  let subgrids = Array.from({ length: 9 }, () => new Set());
  for (let i = 0; i < 9; i++) {
    for (let j = 0; j < 9; j++) {
      const cell = board[i][j];
      if (cell === ".") continue;
      const subgridIndex = Math.floor(i / 3) * 3 + Math.floor(j / 3);
      if (rows[i].has(cell) || cols[j].has(cell) || subgrids[subgridIndex].has(cell)) {
        return false;
      }
      rows[i].add(cell);
      cols[j].add(cell);
      subgrids[subgridIndex].add(cell);
    }
  }
  return true;
}
export {};
